File: kp_algo.go
package main

import (
	"cmath"
	"container/vector"
	"container/heap"
)

type Bin struct {
	Rect
	top float64
	t   int
}

type HeapElemLike interface {
	VacantSpace() float64
}

type HeapElem struct {
	p *Bin
}

func (v *HeapElem) VacantSpace() float64 {
	return v.p.h - v.p.top
}

func (v *HeapElem) Less(x interface{}) bool {
	return v.VacantSpace() > x.(HeapElemLike).VacantSpace()
}

// Packs rectangles from slice 'rects' to the strip starting from lower bound 
// 'be' according to Kuzyurin-Pospelov's basic algorithm. 
// Returns an upper bound of resulting alignment. This algo in not quite 
// on-line, it uses number of all the rectangles.
// Ignores 'm' parameter since it is single strip packing algorithm.
type Kp1Algo struct {
	frame    Bin
	bins     map[int]*vector.Vector
	delta, u float64
	d        int
}

func (v *Kp1Algo) Init(n int) {
	v.delta = real(cmath.Pow(complex(float64(n), 0), (-1.0 / 3)))
	v.u = real(cmath.Pow(complex(float64(n), 0), (1.0 / 3)))
	v.d = int(1 / (2 * v.delta))
	v.bins = make(map[int]*vector.Vector)
	for y := 0; y <= 2*v.d+1; y++ {
		vec := make(vector.Vector, 0)
		v.bins[y] = &vec
	}
}

func (v *Kp1Algo) Pack(rects []Rect, xbe, ybe float64, m int) float64 {
	v.frame.top = 0
	v.frame.y = ybe
	v.frame.x = xbe
	v.frame.w = 1

	n := len(rects)
	v.Init(n)

	for i := 0; i < n; i++ {
		r := &rects[i]
		if r.w > (1 - v.delta) {
			PackToBin(&v.frame, r)
			continue
		}
		
		j := v.RectType(r)
		if v.PackToTopBin(r, j) {
			continue
		}
		// Opening pair of new bins and packing current rectangle into corresponging
		// one.
		v.PackToNewShelfInFrame(r, &v.frame, j)
	}
	return v.frame.y + v.frame.top
}

// Returns true/false whether rectangle was packed into top-of-the-heap bin.
func (v *Kp1Algo) PackToTopBin(r *Rect, j int) bool {
	if 0 == len(*v.bins[j]) {
		return false
	}
	he := heap.Pop(v.bins[j]).(*HeapElem)
	defer heap.Push(v.bins[j], he)
	if he.VacantSpace() >= r.h {
		PackToBin(he.p, r)
		return true
	}
	return false
}

func (v *Kp1Algo) RectType(r *Rect) int {
	j := int(0)
	for y := 1; y <= v.d; y++ {
		if r.w <= (v.delta * float64(y)) {
			j = y
			break
		}
	}
	if 0 == j {
		for y := v.d; y >= 1; y-- {
			if r.w <= (1 - v.delta*float64(y)) {
				j = v.d*2 - y + 1
				break
			}
		}
	}
	return j
}

func (v *Kp1Algo) PackToNewShelfInFrame(r *Rect, f *Bin, j int) {
	b1 := v.AddBin(j)
	PackToBin(f, &b1.Rect)
	PackToBin(b1, r)
	b2 := v.AddBin(v.ComplType(j))
	b2.y = b1.y
	b2.x = b1.x + b1.w
}

func (v *Kp1Algo) AddBin(t int) *Bin {
	b := new(Bin)
	b.h = v.u
	b.w = v.WidthType(t)
	b.top = 0
	b.t = t
	heap.Push(v.bins[t], &HeapElem{b})
	return b
}

func PackToBin(bin *Bin, r *Rect) {
	r.x = bin.x
	r.y = bin.y + bin.top
	bin.top += r.h
}

func (v *Kp1Algo) ComplType(t int) int {
	return 2*v.d - t + 1
}

func (v *Kp1Algo) WidthType(t int) float64 {
	if t <= v.d {
		return v.delta * float64(t)
	}
	return 1 - float64(v.ComplType(t))*v.delta
}

// Algorithm on top of Kp1Algo. Kp2 applies Kp1 for consequtive subsequences of 
// length 2, 4, 8, 16, etc in on-line manner so that Kp2Algo actually does not 
// know  total number of rectangles.
type Kp2Algo struct{}

func (v *Kp2Algo) Pack(rects []Rect, xbe, ybe float64, m int) float64 {
	var b []Rect
	a := rects[:]
	exit_flag := false
	var H float64 = ybe

	for cnt := 2; !exit_flag; cnt *= 2 {
		if cnt > len(a) {
			exit_flag = true
			b = a
		} else {
			b = a[:cnt]
			a = a[cnt:]
		}
		kp1 := new(Kp1Algo)
		H = kp1.Pack(b, xbe, H, m)
	}
	return H
}

File: cc/kp_algo.cc
#include "strip_packing.h"

//void Algorithm::CountRect(Rect* rect) {
 // total_area += rect->w * rect->h;
//}

double Bin::VacantSpace() const {
  return h - top;
}

bool operator< (const Bin& a, const Bin& b) {
  return a.VacantSpace() > b.VacantSpace();
}

double Kp1Algo::Pack(int n, double xbe, double ybe, Context* context) {
  frame_.top = 0;
  frame_.y = ybe;
  frame_.x = xbe;
  frame_.w = 1;
  
  InitParams(n);
  Rect r;
  for (int i = 0; i < n; ++i) {
    NextRect(&r);
    if (r.w > (1 - delta_)) {
      PackToBin(&frame_, &r);
    } else {
      int j = RectType(&r);
      if (!PackToTopBin(&r, j)) {
        PackToNewShelfInFrame(&r, &frame_, j);
      }
    }
    if (context->opt.save_rects) {
      saved_rects.push_back(SavedRect(r, "black", true));
    }
  }
  SaveBins(context);
  solution_height = frame_.y + frame_.top;
  return solution_height;
}

void Kp1Algo::SaveBins(Context* context) {
  if (!context->opt.render_bins) {
    return;
  }
  SetOfBins::iterator i;
  for (int y = 0; y <= d_ * 2 + 1; ++y) {
    for (i = bins_[y].begin(); i != bins_[y].end(); ++i) {
      saved_rects.push_back(SavedRect(*i, "red", false));
    }
  }
}

void Kp1Algo::InitParams(int n) {
  saved_rects.clear();
  bins_.clear();
  delta_ = pow(double(n), -1.0 / 3.0);
  u_ = pow(double(n), 1.0 / 3.0);
  d_ = int(1.0 / (2 * delta_));
}

// Returns true/false whether rectangle was packed into top-of-the-heap bin.
// @j is the type of rectangle @r.
bool Kp1Algo::PackToTopBin(Rect* r, int j) {
  if (0 == bins_[j].size()) {
    return false;
  }
  MapOfSets::iterator mi = bins_.find(j);
  SetOfBins::iterator i = mi->second.begin();
  Bin b = *i;
  if (b.VacantSpace() >= r->h) {
    mi->second.erase(i);
    PackToBin(&b, r);
    mi->second.insert(b);
    return true;
  }
  return false;
}

int Kp1Algo::RectType(Rect* r) {
  for (int y = 1; y <= d_; y++) {
    if (r->w <= (delta_ * double(y))) {
      return y;
    }
  }
  for (int y = d_; y >= 1; --y) {
    if (r->w <= (1 - delta_ * double(y))) {
      return d_ * 2 - y + 1;
    }
  }
}

void Kp1Algo::PackToNewShelfInFrame(Rect* r, Bin* f, int j) {
  int cj = ComplType(j);
  Bin b1(WidthType(j), u_, j, 0);
  Bin b2(WidthType(cj), u_, cj, 0);
  PackToBin(f, &b1);
  PackToBin(&b1, r);
  b2.y = b1.y;
  b2.x = b1.x + b1.w;
  bins_[j].insert(b1);
  bins_[cj].insert(b2);
}

int Kp1Algo::ComplType(int t) {
	return 2 * d_ - t + 1;
}

double Kp1Algo::WidthType(int t) {
	if (t <= d_) {
		return delta_ * double(t);
	}
	return 1.0 - double(ComplType(t)) * delta_;
}